gef.py

```diff
@@ -3257,6 +3304,213 @@ def del_setting(self, name):
+class PieInfoCommand(GenericCommand):
+    """Display breakpoint info"""
+
+    _cmdline_ = "pie info"
+    _syntax_  = "{:s} BREAKPOINT".format(_cmdline_)
+
+    def do_invoke(self, argv):
+        global __pie_breakpoints__
+        if len(argv) < 1:
+            # No breakpoint info needed
+            bps = [__pie_breakpoints__[x] for x in __pie_breakpoints__]
+        else:
+            try:
+                bps = map(
+                    lambda x: __pie_breakpoints__[int(x)],
+                    argv
+                )
+            except ValueError:
+                err("Please give me breakpoint number")
+                return
+        lines = []
+        lines.append('VNum\tNum\tAddr')
+        lines += map(
+                lambda x:'{}\t{}\t{}'.format(x.vbp_num, x.bp_num if x.bp_num else 'N/A', x.addr),
+                bps
+            )
+        print('\n'.join(lines))
+
+
+@register_command
+class PieDeleteCommand(GenericCommand):
+    """Delete a PIE breakpoint"""
+
+    _cmdline_ = "pie delete"
+    _syntax_  = "{:s} [BREAKPOINT]".format(_cmdline_)
+
+    def do_invoke(self, argv):
+        global __pie_breakpoints__
+        if len(argv) < 1:
+            # no arg, delete all
+            to_delete = [__pie_breakpoints__[x] for x in __pie_breakpoints__]
+            self.delete_bp(to_delete)
+        try:
+            self.delete_bp(map(lambda x: __pie_breakpoints__[int(x)], argv))
+        except ValueError:
+            err('Please input PIE virtual breakpoint number to delete')
+
+    @staticmethod
+    def delete_bp(breakpoints):
+        global __pie_breakpoints__
+        for bp in breakpoints:
+            # delete current real breakpoints if exists
+            gdb.execute(
+                'delete {}'.format(bp.bp_num)
+            ) if bp.bp_num else 0
+            # delete virtual breakpoints
+            del __pie_breakpoints__[bp.vbp_num]
+
+
+@register_command
+class PieRunCommand(GenericCommand):
+    """Run process with PIE breakpoint support"""
+
+    _cmdline_ = "pie run"
+    _syntax_  = _cmdline_
+
+    def do_invoke(self, argv):
+        global __pie_breakpoints__
+        fpath = get_filepath()
+        if fpath is None:
+            warn("No executable to debug, use `file` to load a binary")
+            return
+
+        if not os.access(fpath, os.X_OK):
+            warn("The file '{}' is not executable.".format(fpath))
+            return
+
+        if is_alive():
+            warn("gdb is already running. Restart process.")
+
+        # get base address
+        gdb.execute("set stop-on-solib-events 1")
+        disable_context()
+        gdb.execute("run")
+        enable_context()
+        gdb.execute("set stop-on-solib-events 0")
+        vmmap = get_process_maps()
+        base_address = [x.page_start for x in vmmap if x.path == get_filepath()][0]
+        info('base address {}'.format(hex(base_address)))
+
+        # modify all breakpoints
+        for bp in __pie_breakpoints__:
```

Comment:
use `for key, value in __pie_breakpoints__.items():` to avoid the extra lookup below.

Reply:
I'll fix them right away. I haven't considered so much while implementing this. Thanks for reviewing. I s this idea good? By adding a series of commands to support PIE?